subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java

```diff
@@ -493,27 +495,65 @@ default void fromDependencyCollector(DependencyCollector collector) {
+     *
      * <ul>
      *     <li>The {@link Configuration} itself is resolved</li>
-     *     <li>Another {@link Configuration} that extends this one is resolved</li>
-     *     <li>Another {@link Configuration} that references this one as a project dependency is resolved</li>
+     *     <li>The {@link Configuration} is published to a repository</li>
+     *     <li>The {@link Configuration} is consumed as a variant by another project</li>
+     *     <li>Another {@link Configuration} that extends this one is resolved, published, or consumed</li>
      * </ul>
+     * <p>
+     * In general, this method should be avoided in favor of other lazy APIs. However, in some cases where
+     * lazy APIs are not yet available, this method can be used to perform actions before the configuration
+     * is used.
+     * <p>
+     * Despite the method's name, callbacks registered on this method should not add dependencies to the
+     * configuration or mutate the dependencies already present on the configuration. Instead, use the
+     * {@link Provider}-accepting methods on {@link DependencySet} and {@link DependencyConstraintSet}.
+     * <p>
+     * Consider the following example that lazily adds a dependency to a configuration:
      *
-     * This method is useful for mutating the dependencies for a configuration:
      * <pre class='autoTested'>
      * configurations { conf }
-     * configurations['conf'].withDependencies { dependencies -&gt;
-     *      dependencies.each { dependency -&gt;
-     *          if (dependency.version == null) {
-     *              dependency.version { require '1.0' }
-     *          }
-     *      }
+     * configurations['conf'].dependencies.addLater(provider {
+     *     project.dependencies.create("com:example:1.0")
+     * })
+     * </pre>
+     *
+     * Similarly, instead of mutating an existing dependency, use dependency constraints instead. Consider the following
+     * example that uses a dependency constraint to prefer a specific version of a dependency if the
+     * dependency has not declared a preferred version itself. If the dependency has declared a version, the preferred
+     * version constraint will be ignored:
+     *
+     * <pre class='autoTested'>
+     * configurations { conf }
+     * dependencies {
+     *     conf("com:example")
+     *
+     *     constraints {
+     *         conf("com:example") {
+     *             version {
+     *                 prefer("2.0")
+     *             }
+     *         }
+     *     }
      * }
      * </pre>
      *
-     * Actions will be executed in the order provided.
+     * In some cases, using this method may still be necessary:
+     *
+     * <ul>
+     *     <li>
+     *         <strong>Adding excludes</strong>: This method may be used to lazily add excludes to a Configuration.
+     *         Adding excludes to declared dependencies should be handled with {@link ComponentMetadataHandler component metadata rules}
```

Comment:
Small recommendation so it reads slightly better:
```suggestion
     *         Adding excludes to declared dependencies should be handled with {@linkplain ComponentMetadataHandler component metadata rules}
```